sql/analyzer/apply_join.go

```diff
@@ -245,3 +246,153 @@ func normalizeSelectSingleRel(ctx *sql.Context, a *Analyzer, n sql.Node, scope *
 		return c.Node, transform.SameTree, nil
 	})
 }
+
+// hoistOutOfScopeFilters pulls filters upwards into the parent scope
+// to decorrelate subqueries for further optimizations.
+//
+// select * from xy where exists (select * from uv where x = 1)
+// =>
+// select * from xy where x = 1 and exists (select * from uv)
+func hoistOutOfScopeFilters(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope, sel RuleSelector) (sql.Node, transform.TreeIdentity, error) {
+	switch n.(type) {
+	case *plan.TriggerBeginEndBlock:
+		return n, transform.SameTree, nil
+	}
+	ret, same, _, err := recurseSubqueryForOuterFilters(n, a, scope)
```

Comment:
Looks like we don't ever expect this top level call to return any filters to hoist up, but I'm wondering if it's still worth a check that the returned hoistFilters is empty, and returning an error if not.

Reply:
added